Adding GUFunc order, dtype and casting kwarg support

emcastillo opened the pull request:
This mostly completes the support for GUFunc adding the same core characteristics than numpy

asi1024 commented on cupy/_core/_gufuncs.py:
```diff
@@ -141,6 +142,167 @@ def _validate_normalize_axes(
+            ops.append(_OpsRegister._Op(ins, outs, op))
+        return ops
+
+    def _can_cast(self, d1, d2, casting):
```
In what case does this function behave differently from `numpy.can_cast` itself?

emcastillo replied:
Its the same but with a shortcut, this was taken from kernel.pyx iirc

emcastillo commented on cupy/_core/_gufuncs.py:
```diff
@@ -141,6 +142,167 @@ def _validate_normalize_axes(
+    def _sig_str_to_tuple(self, sig):
+        sig = sig.replace(' ', '')
+        toks = sig.split('->')
+        if sig.startswith('->'):
+            ins = ''
+            outs = toks
+        elif sig.endswith('->'):
+            ins = toks
+            outs = ''
```
yeah, sorry about m stupidity here

emcastillo commented:
Added tests for both order and signatures and fixed some quirks with # of outputs.

PTAL!

I will add tests for the rest of features (I did it before but launched a git clean before adding the file :() and expose gufunc to the user via cupyx in a different PR.

asi1024 commented:
Jenkins, test this please.

emcastillo commented:
failure is not related :)

asi1024 commented:
I confirmed that `test_gufunc.py` ran without errors in windows CI. LGTM.

emcastillo commented:
Thank you!
